flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserBaseSemanticAnalyzer.java

```diff
@@ -1862,6 +1868,70 @@ public static RelNode genValues(
                 rows);
     }
 
+    /**
+     * traverse the given node to find all correlated variables, the main logic is from {@link
+     * HiveFilter#getVariablesSet()}.
+     */
+    public static Set<CorrelationId> getVariablesSetForFilter(RexNode rexNode) {
```

Comment:
After reading the code in `SqlToRelConverter#getCorrelationUse`, I find the code is buggy here. We should only add the correlation id on the filter when we can resolve it in the current scope. But the code here will add the correlation id if it is a subquery and its condition contains the correlated variables.  In Calcite, for the query
```
select * from x x1 
where x1.a in (
  select x2.a from x x2
  where exists (
     select x3.a from x x3 where x3.a = x1.a
  )
);
```

has the RelNode tree:
```
LogicalProject(inputs=[0..1])
+- LogicalFilter(condition=[IN($0, {
LogicalProject(a=[$0])
  LogicalFilter(condition=[EXISTS({
LogicalFilter(condition=[=($0, $cor0.a)])
  LogicalTableScan(table=[[default_catalog, default_database, x, source: [TestTableSource(a, b)]]])
})])
    LogicalTableScan(table=[[default_catalog, default_database, x, source: [TestTableSource(a, b)]]])
})], variablesSet=[[$cor0]])
   +- LogicalTableScan(table=[[default_catalog, default_database, x, source: [TestTableSource(a, b)]]])
```
where only the out-most LogicalFilter contains correlation id.

Considering we just align with the Hive behavior, it's fine we don't support this.